sw/host/opentitanlib/src/proxy/mod.rs

```diff
@@ -2,4 +2,75 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
+use anyhow::{bail, Result};
+use handler::TransportCommandHandler;
+use protocol::Message;
+use socket_server::JsonSocketServer;
+
+use crate::app::TransportWrapper;
+
+mod handler;
 pub mod protocol;
+mod socket_server;
+
+/// Interface for handlers of protocol messages, responding to each message with a single
+/// instance of the same protocol message.
+pub trait CommandHandler<Msg> {
+    fn execute_cmd(&self, msg: &Msg) -> Result<Msg>;
+}
+
+/// This is the main entry point for the session proxy.  This struct will either bind on a
+/// specified port, or find an available port from a range, before entering an event loop.
+pub struct SessionHandler<'a> {
+    port: u16,
+    socket_server: JsonSocketServer<Message, TransportCommandHandler<'a>>,
+}
+
+impl<'a> SessionHandler<'a> {
+    pub fn init(transport: &'a TransportWrapper, listen_port: &Option<u16>) -> Result<Self> {
```

Comment:
Why take a reference of a Copy type?

Reply:
Hm, I thought of Option<> as a "complicated" object, which could be expensive to copy, but thinking about it, it would probably be represented as a single 32-bit value, easily being passed in a register.